### ISSUE TITLE
Http Server as windows service

### ISSUE THREAD

**Issue by charliemaiors** [kardianos/service#108]
Hi, I've defined simple service which stars an HTTP server I've defined the service installation [here](https://github.com/charliemaiors/sleep-on-lan/blob/master/server/service.go) but when I try to start service from PowerShell or simple services.msc it will return the error 1053: the service did not respond to the start or control request in a timely fashion.

**Comment by mahaben** [comment]
@charliemaiors  I'm facing the same issue. Did you find some workaround? Thanks

**Comment by charliemaiors** [comment]
Yes I'ive created a Powershell [script](https://github.com/charliemaiors/sleep-on-lan/blob/master/script/win-service-install.ps1) which install [nssm](https://nssm.cc/) from chocolatey and define the service in nssm.

**Comment by mahaben** [comment]
I was using nssm previously and I don't want to use it anymore. I'll try to find out a solution. Thank you anyway.

**Comment by bbalet** [comment]
For Windows, the main function must be the fastest possible so as to return very quickly a kind of ack to the OS. So you shouldn't do some many operations such as initialising things but be as brief as possible. Loading configuration, initialising stuffs and doing the job should be delegated to another function that is the service. See an example here (in French, but the sample code is commented on English): http://decouvric.cluster013.ovh.net/golang/thirdparty/divers/creer-un-service-golang-avec-kardianos.html

**Comment by mahaben** [comment]
Thanks. But I can't understand when the function s.Run is getting called?

**Comment by bbalet** [comment]
The part of code where s.Run is used is the service itself. As you can see, s.Run is called very quickly (nothing big is initialised before the call and no parameter is given to a Windows service). This is the pattern you must follow if you want to code a Windows service.

**Comment by nkev** [comment]
My HTTP is unreachable when running under Windows service, even via localhost. Works fine if not running under Windows service. My s.Run is fast and and there is no error that I can see in the Windows Event logs. Has anybody encountered this?

**Comment by bbalet** [comment]
Windows Firewall can block the network port used by your service, even when accessed from localhost to localhost. Try to disable it, if it is the root cause, allow the port.
An antivirus could be the root cause too.

**Comment by bbalet** [comment]
When your program is executed as a stand alone process, it doesn't use the same account than when it runs as a service (so the level of privileges and the context differ).

**Comment by PauloSalum** [comment]
Has anyone found a solution to this? I'm having the same issue.

**Comment by PauloSalum** [comment]
> I was using nssm previously and I don't want to use it anymore. I'll try to find out a solution. Thank you anyway.

@mahaben did you find a solution for this?  I'm having the same issue.

**Comment by kardianos** [comment]
Your start function must return in milliseconds.
If in doubt post code.

**Comment by Sozialarchiv** [comment]
Hi I have the same problem. If there is nothing to do everything work. As soon as I add an http server I get the 1053 error:

```
package main

import (
	"fmt"
	"github.com/kardianos/service"
	"log"
	"ims/webserver"
	"time"
)

var logger service.Logger

type program struct{}

func (p *program) Start(s service.Service) error {
	// Start should not block. Do the actual work async.
	go p.run()
	return nil
}
func (p *program) run() {
	fmt.Println(time.Now().Format("2006-01-02 03:04:05 PM"), "Service started")
	webserver.StartWebserver()
}
func (p *program) Stop(s service.Service) error {
	// Stop should not block. Return with a few seconds.
	return nil
}

func main() {
	svcConfig := &service.Config{
		Name:        "ims",
		DisplayName: "IMS",
		Description: "IMS",
	}

	prg := &program{}
	s, err := service.New(prg, svcConfig)
	if err != nil {
		log.Fatal(err)
	}
	logger, err = s.Logger(nil)
	if err != nil {
		log.Fatal(err)
	}
	err = s.Run()
	if err != nil {
		logger.Error(err)
	}
}
```

Thank you for your help.

**Comment by kardianos** [comment]
Maybe "webserver.StartWebserver()" is panicing? Does it work if you comment that out?

**Comment by Sozialarchiv** [comment]
> Maybe "webserver.StartWebserver()" is panicing? Does it work if you comment that out?

You are right. It was a panic, that was just occuring if it was running as service. I solved the problem and now everthing works great.

Thank you for your fast answer and your great package!

**Comment by mcauto** [comment]
It seems that when running a server as a Windows service, the network resources used by the process are automatically released upon termination.
(A similar behavior occurs when a client connects to the server as well.)

Has anyone encountered and resolved this issue?
